Firestore/core/src/firebase/firestore/remote/serializer.h

```diff
@@ -17,27 +17,109 @@
 #ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_SERIALIZER_H_
 #define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_SERIALIZER_H_
 
+#include "Firestore/Protos/nanopb/google/firestore/v1beta1/document.pb.h"
+#include "Firestore/core/src/firebase/firestore/model/database_id.h"
+#include "Firestore/core/src/firebase/firestore/model/field_value.h"
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+
 namespace firebase {
 namespace firestore {
 namespace remote {
 
 /**
  * @brief Converts internal model objects to their equivalent protocol buffer
- * form.
+ * form, and protocol buffer objects to their equivalent bytes.
  *
- * Methods starting with "Encode" convert to a protocol buffer and methods
- * starting with "Decode" convert from a protocol buffer.
+ * Methods starting with "Encode" either convert from a model object to a
+ * protocol buffer or from a protocol buffer to bytes, and methods starting with
+ * "Decode" either convert from a protocol buffer to a model object or from
+ * bytes to a protocol buffer.
  *
  */
 // TODO(rsgowman): Original docs also has this: "Throws an exception if a
 // protocol buffer is missing a critical field or has a value we can't
 // interpret." Adjust for C++.
 class Serializer {
  public:
-  // TODO(rsgowman): Adjust ctor to accept a DatabaseID... once that exists.
-  Serializer(/*DatabaseID databaseId*/);
+  /**
+   * @brief Wraps (nanopb) google_firestore_v1beta1_Value with type information.
+   */
+  struct ValueWithType {
+    firebase::firestore::model::FieldValue::Type type;
+    google_firestore_v1beta1_Value value;
+  };
+
+  explicit Serializer(firebase::firestore::model::DatabaseId database_id)
+      : database_id_(database_id) {
+  }
+
+  /**
+   * Converts the FieldValue model passed into the Value proto equivalent.
+   *
+   * @param field_value the model to convert.
+   * @return the proto representation of the model.
+   */
+  static Serializer::ValueWithType EncodeFieldValue(
+      const firebase::firestore::model::FieldValue& field_value);
+
+  /**
+   * @brief Converts the value proto passed into bytes.
+   *
+   * @param[out] out_bytes A buffer to place the output.
+   * @param[in,out] inout_bytes_length On input, the length of the out_bytes
+   * buffer. On output, will contain the number of bytes actually used.
+   */
+  // TODO(rsgowman): error handling, incl return code.
+  // TODO(rsgowman): how large should the output buffer be? The docs on this
+  // method give no indication. (Unfortunately, neither does the underlying
+  // nanopb calls; we may be stuck with something vague.) But we can do some
+  // further investigation to see if there's some way to get nanopb to tell us
+  // how much space it's going to need and then expose/document that here.
+  static void EncodeValueWithType(const ValueWithType& value,
+                                  uint8_t* out_bytes,
```

Comment:
> Very optional: perhaps absl::Span could be used here. It's a string_view-like non-owning wrapper that holds pointer to array and the size together. Then this function could return the number of bytes actually written and avoid an input/output parameter (though if you plan to return an error code, then this doesn't apply).

I'm planning to add a return code (as soon as I get around to handling errors.) That said, I don't object to using Span anyway, though we don't currently have it in our code base, so I'm reluctant to add it just for this.

> Closely related, I'm kinda wary about the input/output parameter. Does the caller always care about the number written? 

I would think so. It'll be the code that interacts with grpc, and while I'm not sure how that works yet, at some point, it'll need to know how many bytes to write to the network.

> If not, then maybe something like (const ValueWithType& value, uint8_t* out_bytes, size_t size, size_t* bytes_written = nullptr)?

> Also, it makes sense for Nanopb to leave the allocation of the buffer to the caller. Does the same apply to the Serializer, or perhaps it could take care of the allocation strategy as well (assuming there is allocation involved. Maybe the callers use static buffers)? If that were the case, then maybe the signature could hide the low-level details of communication with Nanopb and accept/return some easier-to-use type (say, vector<char>/unique_ptr<char[]>)?

Again, I'm not sure just yet, but my assumption is that something that interfaces with grpc will be the caller here. It wouldn't be unreasonable for grpc to simply have a static buffer that it reuses... but I simply don't know enough about how that works just yet.

We've already got a TODO there related to allocation, so I'll add your suggestion to it, but will otherwise leave this as is for now until we have a better idea on how it'll be used.

Reply:
I'm not a huge fan of raw buffers everywhere. I'm particularly not a fan of needing to precompute the size of the buffer or fail if not large enough. Better to make the buffer dynamic and allow each method to grow the buffer as needed.

To answer some of the other questions: gRPC does not assume a static buffer nor does it particularly go out of its way to make that possible. In typically usage with protobuf the buffer is malloced for each RPC.

Ultimately in the C core, the buffer is attached to the `send_message` of the `grpc_op`, which is here:
https://github.com/grpc/grpc/blob/master/include/grpc/impl/codegen/grpc_types.h#L528

gRPC expects to own (and destroy) that buffer. The Objective-C wrapper serializes into an `NSData` and copies that into a buffer it can hand off into gRPC:
https://github.com/grpc/grpc/blob/86accb1768cc860a7579bb5fff9bd3c2ba7274af/src/objective-c/GRPCClient/private/NSData%2BGRPC.m#L82

I'd suggest simplifying and just doing a vector and appending as we go. If profiling turns up that we're spending a lot of time copying buffers then we can undertake the work of using slices and buffers optimally. However, I'll be _shocked_ if `memcpy` ever becomes out bottleneck.

Reply:
Alright; lets use a vector for now. (Note that we still have the same problem of precomputing the size of the buffer when dealing with nanopb... but that problem is now buried in the implementation details of this class.)